core/src/services/s3/backend.rs

```diff
@@ -517,6 +518,60 @@ impl S3Builder {
 
         self
     }
+
+    /// a helper function to make it easier to find region
+    pub async fn detect_region(&self) -> Result<Option<String>> {
```

Comment:
I feel like we should not read any value from builder which is confusing and hard to use. We can have an API like:

```
pub async fn detect_region(endpoint: &str, bucket: &str) -> Option<String> {
```